### PR TITLE
Harvest: Handle Errors during account, files and trigger mutations

### PR DESCRIPTION
**Pull request by gregorylegarec**
The goal of this PR is to catch and handle errors from konnector job, and regular mutations errors as well.

It starts with a few refactoring:
* Code/test simplifications
* AccountCreator and AccountEditor logics are now too close as AccountCreator needs to keep the created account in case of error, so are both merged into TriggerManager which now handle account creation and update.
* TriggerSuccessMessage disappears, as it is actually more the Cozy-Home's role to display the success message and picture.

Then we create:
* KonnectorJobError class, as we need a dedicated logic to handle errors from konnector jobs
* AccountFormError component, which display Errors and KonnectorJobErrors

At last, we catch errors in TriggerManager and handle them in AccountForm.

NB: During my last tests I realized that a case is not well covered. When a directory creation, permission creation, or reference addition fails, the account is now orphan and any update cannot launch the associated trigger. I will amend this PR with a fix that check directory and trigger existence and will create them if they does not exist. But this PR can yet be reviewed.

### CODE REVIEW

File: packages/cozy-harvest-lib/test/components/AccountForm.spec.js [hunk 1]
```diff
@@ -101,39 +101,71 @@ describe('AccountForm', () => {
...
-  it('should have disabled button if there is required field empty', () => {
-    const fields = {
-      test: {
-        type: 'text'
+  describe('Submit Button', () => {
+    const getButtonDisabledValue = wrapper =>
+      wrapper
+        .dive()
+        .find('DefaultButton')
+        .props().disabled
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/test/components/AccountForm.spec.js]: Nit: I think getButtonDisabledValue() does a bit too much stuff (finding the button, and checking a prop), I find it a bit cumbersome to pass wrapper also.

I like to keep a global wrapper variable that helper function use directly.

It is a matter of taste and not blocking at all.

```
let wrapper

const findButton = () => wrapper.dive().find('DefaultButton').props().disabled

it('should do stuff', () => {
  wrapper = shallow(...)
  const btn = findButton()
  expect(btn.props().disabled).toBe(true)
})
```

Why is it necessary to `dive()` ?

File: packages/cozy-harvest-lib/src/components/AccountCreator.jsx [hunk 2]
```diff
@@ -36,10 +36,8 @@ export class AccountCreator extends PureComponent {
     const { account, konnector, submitting } = this.props
     return (
       <AccountForm
-        fields={konnector.fields}
+        konnector={konnector}
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/components/AccountCreator.jsx]: 🎉 

I think it is much simpler mentally to pass around known objects and it avoids creating a lot of props.

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 3]
```diff
@@ -104,57 +96,75 @@ export class TriggerManager extends Component {
...
+    const { account, error, status } = this.state
+    const submitting = status === RUNNING || running
+
+    return (
+      <AccountForm
+        error={error}
         konnector={konnector}
-        onBeforeCreate={this.handleAccountMutation}
-        onCreateSuccess={this.handleAccountCreationSuccess}
+        initialValues={account && (account.auth || account.oauth)}
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: Pass directly the account no ? Why is it ncessary to destructure here ?

File: packages/cozy-harvest-lib/src/connections/accounts.js [hunk 4]
```diff
@@ -109,14 +109,27 @@ const updateAccount = async (client, account) => {
...
 /**
  * Get accounts mutations
  * @param  {Object} client CozyClient
  * @return {Object}        Object containing accounts mutations
  */
 export const accountsMutations = client => ({
   createAccount: createAccount.bind(null, client),
-  updateAccount: updateAccount.bind(null, client)
+  updateAccount: updateAccount.bind(null, client),
+  saveAccount: saveAccount.bind(null, client)
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/connections/accounts.js]: Why is updateAccount necessary here ? saveAccount already does both creating and saving ?

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 5]
```diff
@@ -104,57 +96,75 @@ export class TriggerManager extends Component {
...
-    this.setState({
-      status: SUCCESS
-    })
+    return onSuccess(trigger)
   }
 
   render() {
-    const { account, konnector, onDone, running } = this.props
-    const { createdAccount, status, updatedAccount } = this.state
-    const succeed = createdAccount && [LOGGED, SUCCESS].includes(status)
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: No more LOGGED or SUCCESS status ? Is this done higher up ?

File: packages/cozy-harvest-lib/src/helpers/permissions.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import intersection from 'lodash/intersection'
+
+export const build = (doc, verbs) => ({
+  type: doc._type,
+  values: [doc._id],
+  verbs: intersection(verbs, ['DELETE', 'GET', 'POST', 'PUT'])
+})
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/helpers/permissions.js]: I do not see where this is used ?

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 7]
```diff
@@ -4,9 +4,8 @@ import PropTypes from 'react-proptypes'
 import { withMutations } from 'cozy-client'
 import { translate } from 'cozy-ui/react/I18n'
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: This commit (73178ab190a997dc851a5ac36d11a341d36c9d35) is a bit hard to follow and I am not sure I can properly review it.

File: packages/cozy-harvest-lib/test/components/TriggerManager.spec.js [hunk 8]
```diff
@@ -189,6 +212,52 @@ describe('TriggerManager', () => {
       wrapper.instance().handleSubmit()
       expect(wrapper.props().submitting).toEqual(true)
     })
+
+    it('should call saveAccount without account', () => {
+      const wrapper = shallowWithoutAccount()
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/test/components/TriggerManager.spec.js]: Nit : I find that it would preferable if this function would not two things at once (shallow rendering and passing the right props).

```
const wrapper = shallow(<TriggerManager {...props.noAccount} />)
```

File: packages/cozy-harvest-lib/test/components/TriggerManager.spec.js [hunk 9]
```diff
@@ -189,6 +212,52 @@ describe('TriggerManager', () => {
...
+        fixtures.account
+      )
+    })
+
+    it('should call saveAccount with account', () => {
+      const wrapper = shallowWithAccount()
+      wrapper.instance().handleSubmit(fixtures.data)
+      expect(saveAccountMock).toHaveBeenCalledWith(
+        fixtures.konnector,
+        fixtures.createdAccount
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/test/components/TriggerManager.spec.js]: I find a bit weird that saveAccount is called with first argument a konnector ?

File: packages/cozy-harvest-lib/src/components/AccountCreator.jsx (hunk 2 above)
Comment (gregorylegarec) [reply packages/cozy-harvest-lib/src/components/AccountCreator.jsx]: Totally agree, it was a matter of time :)

File: packages/cozy-harvest-lib/src/connections/accounts.js (hunk 4 above)
Comment (gregorylegarec) [reply packages/cozy-harvest-lib/src/connections/accounts.js]: Yep, I first removed them but then I put them back as it was easier to test them (i.e just keep their related tests).

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx (hunk 5 above)
Comment (gregorylegarec) [reply packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: Nope. The triggerManager is either `IDLE` or `RUNNING`.

The success is handler higher up, via `onLoginSuccess` and `onSuccess` hooks.

File: packages/cozy-harvest-lib/src/helpers/permissions.js (hunk 6 above)
Comment (gregorylegarec) [reply packages/cozy-harvest-lib/src/helpers/permissions.js]: Oh well spotted, it has been commited by mistake. Should go in our future permission model instead. I'll remove this file from the PR right now.

File: packages/cozy-harvest-lib/test/components/TriggerManager.spec.js (hunk 9 above)
Comment (gregorylegarec) [reply packages/cozy-harvest-lib/test/components/TriggerManager.spec.js]: `createAccount` is needing a konnector, to know if it needs to create an aggregator account first.

A workaround could be to pass this information as an options or directly into the account.

But you're right, I think it's a right thing to avoid passing a konnector at this point of our code.

I suggest, if you agree, to keep it for a future (little) refactoring.

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 10]
```diff
@@ -66,29 +66,35 @@ export class TriggerManager extends Component {
...
+          accounts.getLabel(account)
+        )}`
+
+        folder =
+          (await statDirectoryByPath(path)) ||
+          (await createDirectoryByPath(path))
+
+        await addPermission(konnector, konnectors.buildFolderPermission(folder))
+        await addReferencesTo(konnector, [folder])
+      }
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: Could be extracted outside the component in a  ensureKonnectorHasFolderIfNeeded(konnector) method

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 11]
```diff
@@ -66,29 +66,35 @@ export class TriggerManager extends Component {
...
+          cron: cron.fromKonnector(konnector),
+          folder,
+          konnector
+        })
+      )
+
+      return await this.launch(trigger)
+    } catch (error) {
+      return this.handleError(error)
     }
```
Comment (ptbrowne) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: I think this business logic should be outside the React component.

File: packages/cozy-harvest-lib/src/components/AccountFormError.jsx [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+import PropTypes from 'react-proptypes'
+
+import { translate } from 'cozy-ui/react/I18n'
+
+import { KonnectorJobError } from '../helpers/konnectors'
+import Markdown from './Markdown'
+
+/**
+ * Component used to show an error related to account or to konnector job
+ * Regular error are treated in a common way, but for KonnectorJobErrors
```
[the remark this answers is not in the document]
Comment (gregorylegarec) [reply packages/cozy-harvest-lib/src/components/AccountFormError.jsx]: classic ?

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx (hunk 11 above)
Comment (gregorylegarec) [reply packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: We agree, the next step is to start using Models.

File: packages/cozy-harvest-lib/src/components/AccountFormError.jsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (ptbrowne) [reply packages/cozy-harvest-lib/src/components/AccountFormError.jsx]: What do you mean by classic or common ? :D

File: packages/cozy-harvest-lib/src/components/AccountCreator.jsx (hunk 2 above)
Comment (y-lohse) [reply packages/cozy-harvest-lib/src/components/AccountCreator.jsx]: Funny seeing this after the other review :p It's ok if the object is *known*, but for me at least `konnector` is opaque and I have no idea what is or isn't in it, so i actually prefer being explicit about the props. It can also help avoid extra renders when used in PureComponents.

File: packages/cozy-harvest-lib/src/components/AccountForm.jsx [hunk 13]
```diff
@@ -284,9 +286,7 @@ export class AccountForm extends PureComponent {
 
 AccountForm.propTypes = {
   account: PropTypes.object,
-  fields: PropTypes.object,
-  oauth: PropTypes.object,
-  locales: PropTypes.object,
+  konnector: PropTypes.object.isRequired,
```
Comment (y-lohse) [inline packages/cozy-harvest-lib/src/components/AccountForm.jsx]: But also if we're passing a full object, we can describe it's shape here :-)

File: packages/cozy-harvest-lib/src/components/AccountFormError.jsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (ptbrowne) [reply packages/cozy-harvest-lib/src/components/AccountFormError.jsx]: Ah ok it's clearer to me now

File: packages/cozy-harvest-lib/src/components/AccountFormError.jsx [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+    const args = {
+      name: konnector.name || '',
+      link: konnector.vendor_link || ''
+    }
+
+    if (error instanceof KonnectorJobError) {
+      return t(getKonnectorJobKey(error.code), {
+        ...args,
+        _: t(getKonnectorJobKey(error.type), {
```
Comment (y-lohse) [inline packages/cozy-harvest-lib/src/components/AccountFormError.jsx]: What is this `_` key?

File: packages/cozy-harvest-lib/src/components/AccountFormError.jsx [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    return error.message ? `${description} (${error.message})` : description
+  }
+
+  render() {
+    const errorDescription = this.getDescription()
+    /* eslint-disable-next-line no-console */
+    console.error(errorDescription)
+    return (
+      <div className="u-error">
+        <Markdown source={this.getDescription()} />
```
Comment (y-lohse) [inline packages/cozy-harvest-lib/src/components/AccountFormError.jsx]: We can use `errorDescription ` here.

File: packages/cozy-harvest-lib/src/components/AccountForm.jsx [hunk 16]
```diff
@@ -193,8 +201,9 @@ export class AccountForm extends PureComponent {
    * @param  {Object} formState See https://github.com/final-form/final-form#formstate
    * @return {Boolean}
    */
-  isSubmittable({ dirty, initialValues, valid }) {
-    return valid && !(initialValues && !dirty)
+  isSubmittable({ dirty, error, initialValues, valid }) {
+    const untouched = initialValues && !dirty
+    return error || (valid && !untouched)
```
Comment (y-lohse) [inline packages/cozy-harvest-lib/src/components/AccountForm.jsx]: So it's submittable when there is an error?

File: packages/cozy-harvest-lib/src/connections/triggers.js [hunk 17]
```diff
@@ -56,12 +57,15 @@ const waitForLoginSuccess = async (
...
 
-  return new Promise(resolve => {
+  return new Promise((resolve, reject) => {
     setTimeout(() => resolve(job), expectedSuccessDelay)
-    jobSubscription.onUpdate(
-      realtimeJob =>
-        JOB_END_STATES.includes(realtimeJob.state) && resolve(realtimeJob)
-    )
+    jobSubscription.onUpdate(updatedJob => {
+      return (
```
Comment (y-lohse) [inline packages/cozy-harvest-lib/src/connections/triggers.js]: Why are we using `return` here?

File: packages/cozy-harvest-lib/src/components/AccountFormError.jsx (hunk 14 above)
Comment (gregorylegarec) [reply packages/cozy-harvest-lib/src/components/AccountFormError.jsx]: It allows to specify a fallback value is the key does not exist in translation file.
See (Polyglot documentation)[https://github.com/airbnb/polyglot.js?files=1#polyglotprototypetkey-interpolationoptions].

Here we are using this feature on three levels 💪, it spares us [this kind of mess](https://github.com/cozy/cozy-home/blob/master/src/lib/konnectors.js#L99)

File: packages/cozy-harvest-lib/src/components/AccountFormError.jsx (hunk 15 above)
Comment (gregorylegarec) [reply packages/cozy-harvest-lib/src/components/AccountFormError.jsx]: Oops, my bad. Very recent change. Well spotted !

File: packages/cozy-harvest-lib/src/components/AccountForm.jsx (hunk 16 above)
Comment (gregorylegarec) [reply packages/cozy-harvest-lib/src/components/AccountForm.jsx]: Yes, first version was not, but actually, if you have an error somewhere you should be allowed to retry, i.e. to resubmit form.

File: packages/cozy-harvest-lib/src/connections/triggers.js (hunk 17 above)
Comment (gregorylegarec) [reply packages/cozy-harvest-lib/src/connections/triggers.js]: 😱

File: packages/cozy-harvest-lib/src/components/AccountForm.jsx (hunk 13 above)
Comment (gregorylegarec) [reply packages/cozy-harvest-lib/src/components/AccountForm.jsx]: [Same remark](https://github.com/cozy/cozy-libs/pull/275#discussion_r258502959) as @Crash--, you definitely should meet each other.

**Comment by gregorylegarec** [comment]
Thanks guys for your precious feedback !
